### PR TITLE
Refactoring websockets #157

### PR DESCRIPTION
**Pull request by sambhavnoobcoder**
## Description
This PR refactors the `AssetsIdentifiersWS` class to inherit from the `BaseWebsocket` class, improving code reusability and standardization across our websocket implementations. The changes maintain the original functionality while adhering to the abstract base class structure.
This Pr addresses the issue #157 

## Detailed Changes

1. Class Inheritance:
   - `AssetsIdentifiersWS` now inherits from `BaseWebsocket`
   - Implemented the abstract `url` property required by `BaseWebsocket`

2. Singleton Pattern:
   - Maintained the existing `__new__` method to ensure `AssetsIdentifiersWS` remains a singleton

3. Initialization and Threading:
   - Kept the original `__init__` method, including the creation of a new event loop and thread
   - Added a call to `super().__init__()` to properly initialize the base class

4. WebSocket Connection:
   - Retained the `open_websocket` method for creating and running the WebSocket connection
   - Updated to use `self.url` instead of directly accessing `Settings.ASSETS_IDENTIFIERS_WS_URL`

5. Message Handling:
   - Adjusted the `on_message` method to pass the `StreamedIdentifiers` object directly to the callback
   - Removed assumptions about the internal structure of `StreamedIdentifiers`

6. Error and Closure Handling:
   - Maintained existing `on_error` and `on_close` methods
   - These methods now properly implement the abstract methods from `BaseWebsocket`

## Testing

- Tested the refactored AssetsIdentifiersWS class to ensure it connects and receives messages correctly
- Verified that the singleton pattern still works as expected
- Confirmed that the threading approach continues to provide non-blocking operations

## Impact of the PR

1. Code Standardization:

- This refactor aligns AssetsIdentifiersWS with other WebSocket implementations in the project
- Establishes a consistent pattern for future WebSocket classes

2. Improved Maintainability:

- Common WebSocket functionality is now centralized in the BaseWebsocket class
- Reduces duplication across different WebSocket implementations

3. Enhanced Extensibility:

- New features added to BaseWebsocket will automatically benefit AssetsIdentifiersWS
- Makes it easier to create new WebSocket classes with shared functionality

4. Potential for Better Error Handling:

- By adhering to the BaseWebsocket structure, we can implement project-wide error handling strategies

5. Minimal Disruption:

- The changes preserve the existing functionality of AssetsIdentifiersWS
- The singleton pattern and threading approach are maintained

7. Clearer Separation of Concerns:

- The url property is now explicitly defined, improving clarity and making it easier to modify if needed

8. Flexibility in Message Handling:

- The updated on_message method is more flexible, not making assumptions about the StreamedIdentifiers structure

### CODE REVIEW

File: src/pieces/assets/assets_identifiers_ws.py [hunk 1]
```diff
@@ -3,31 +3,33 @@
...
-class AssetsIdentifiersWS:
-    def __new__(cls,*args,**kwargs):
+class AssetsIdentifiersWS(BaseWebsocket):
+    def __new__(cls, *args, **kwargs):
         if not hasattr(cls, 'instance'):
             cls.instance = super(AssetsIdentifiersWS, cls).__new__(cls)
         return cls.instance
-    
+
     def __init__(self, on_message_callback):
```
Comment (Bishoy-at-pieces) [inline src/pieces/assets/assets_identifiers_ws.py:17]: Can we remove the init?
We should start the ws in the cli loop so we can change the cli loop file to do BaseWebsocket.start_all

Comment (sambhavnoobcoder) [reply src/pieces/assets/assets_identifiers_ws.py]: addressed in [efad189](https://github.com/pieces-app/cli-agent/pull/167/commits/efad1899ee832844715214fb6d953f4e60315851) commit

Comment (sambhavnoobcoder) [reply src/pieces/assets/assets_identifiers_ws.py]: Also @Bishoy-at-pieces , if we are to use the start_all from the BaseWebsocket eventually , should we also remove the open_websocket and the close_websocket_connection also , as they would then become redundant to the start_all and close_all respectively ?

Comment (Bishoy-at-pieces) [reply src/pieces/assets/assets_identifiers_ws.py]: I would like to leave them 
Later we might need to connect to specific websockets not all for now we might need to start all

Comment (sambhavnoobcoder) [reply src/pieces/assets/assets_identifiers_ws.py]: okay.

**Comment by Bishoy-at-pieces** [comment]
Can we refactor the ask ws also?

**Comment by sambhavnoobcoder** [comment]
I couldn't find the ask ws you were referring to . Could you reference the code you were suggesting for the change ?

**Comment by Bishoy-at-pieces** [comment]
src/pieces/copilot/pieces_ask_websocket.py

**Comment by sambhavnoobcoder** [comment]
hey @Bishoy-at-pieces , I've made the suggested changes to the ask ws as well , kindly give it a look and inform me if any changes are required .

File: src/pieces/copilot/pieces_ask_websocket.py [hunk 2]
```diff
@@ -53,32 +50,27 @@ def on_message(self,ws, message):
...
+        print(f"AskWebsocket closed. Status code: {close_status_code}, Message: {close_msg}")
+        super().on_close(ws, close_status_code, close_msg)
 
     def on_open(self, ws):
-        """Handle websocket opening."""
-        # if self.verbose:
-        #     print("WebSocket connection opened.")
-        self.is_connected = True
-        self.open_event.set()
+        print("AskWebsocket connection opened")
```
Comment (Bishoy-at-pieces) [inline src/pieces/copilot/pieces_ask_websocket.py:72]: We don't really need this

File: src/pieces/copilot/pieces_ask_websocket.py [hunk 3]
```diff
@@ -53,32 +50,27 @@ def on_message(self,ws, message):
...
-        self.is_connected = False
+        print(f"AskWebsocket Error: {error}")
+        super().on_error(ws, error)
 
     def on_close(self, ws, close_status_code, close_msg):
-        """Handle websocket closure."""
-        # if self.verbose:
-        #     print("WebSocket closed")
-        self.is_connected = False
+        print(f"AskWebsocket closed. Status code: {close_status_code}, Message: {close_msg}")
```
Comment (Bishoy-at-pieces) [inline src/pieces/copilot/pieces_ask_websocket.py:68]: No need for this too

File: src/pieces/copilot/pieces_ask_websocket.py [hunk 4]
```diff
@@ -53,32 +50,27 @@ def on_message(self,ws, message):
...
     def on_open(self, ws):
-        """Handle websocket opening."""
-        # if self.verbose:
-        #     print("WebSocket connection opened.")
-        self.is_connected = True
-        self.open_event.set()
+        print("AskWebsocket connection opened")
+        super().on_open(ws)
 
     def _start_ws(self):
```
Comment (Bishoy-at-pieces) [inline src/pieces/copilot/pieces_ask_websocket.py:75]: We should remove this not useful

File: src/pieces/copilot/pieces_ask_websocket.py [hunk 5]
```diff
@@ -1,48 +1,45 @@
...
         self.verbose = True
         self.live = Live()
-        self.message_compeleted = threading.Event()
-    
+        self.message_completed = threading.Event()
+
+    @property
+    def url(self):
+        return Settings.ASK_WEBSOCKET_URL
+
```
Comment (Bishoy-at-pieces) [inline src/pieces/copilot/pieces_ask_websocket.py:29]: We don't need this also we should over write the start method

**Review by Bishoy-at-pieces — CHANGES_REQUESTED** [review]

**Comment by Bishoy-at-pieces** [comment]
Looks super great some small stuff need to be changed but really really great work!

File: base_websocket.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+import websocket
+import threading
+from abc import ABC, abstractmethod
+from typing import Optional, List, Callable
+
+class BaseWebsocket(ABC):
+    instances: List['BaseWebsocket'] = []
```
[the remark this answers is not in the document]
Comment (sambhavnoobcoder) [reply base_websocket.py]: addressed in [324ecb1](https://github.com/pieces-app/cli-agent/pull/167/commits/324ecb1828d95213dedac2b0f7f6fa74add35775) commit

File: src/pieces/copilot/pieces_ask_websocket.py (hunk 3 above)
Comment (sambhavnoobcoder) [reply src/pieces/copilot/pieces_ask_websocket.py]: addressed in commit [73566e9](https://github.com/pieces-app/cli-agent/pull/167/commits/73566e9e9375409f4ec01b4207affc4ea7cec9f2) commit

**Pull request closed without merging**